llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -266,6 +269,103 @@ bool combineINSERT_SUBREG(MachineBasicBlock &MBB) {
   return Changed;
 }
 
+/// This optimization prevents the coalescing and copy propagation of
+/// Addressing registers in loops with shared descriptors for load and stores.
+//  This is important to avoid unnecessary COPYs of such descriptors when
+//  they are coalesced.
+bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
+
+  const TargetInstrInfo &TII = *MBB.getParent()->getSubtarget().getInstrInfo();
+  const AIE2InstrInfo *AII = static_cast<const AIE2InstrInfo *>(&TII);
+  bool Changed = false;
+
+  // TODO: LoopUtils?
+  auto IsLoop = [](MachineBasicBlock *MBB) {
+    int NumLoopEdges = 0;
+    int NumExitEdges = 0;
+    for (auto *S : MBB->successors())
+      if (S == MBB)
+        NumLoopEdges++;
+      else
+        NumExitEdges++;
+    return NumLoopEdges == 1 && NumExitEdges == 1;
+  };
+
+  // We consider only loops here.
+  if (!IsLoop(&MBB))
+    return false;
+
+  DenseMap<Register, SmallPtrSet<MachineInstr *, 8>> LoadUses;
+  SmallSet<Register, 8> UsedByStore;
+
+  // The first part, collect interesting cases.
+  for (MachineInstr &MI : MBB) {
+    if (MI.getOpcode() != TargetOpcode::REG_SEQUENCE)
+      continue;
+
+    const Register DstReg = MI.getOperand(0).getReg();
+    const TargetRegisterClass *RC = MRI.getRegClass(DstReg);
+
+    if (RC != &AIE2::eDRegClass && RC != &AIE2::eDSRegClass)
+      continue;
+
+    // Take first-use instruction, that will be only one due to the
+    // way we lower addressing register's use:
+    // A  =  REG_SEQUENCE __
+    // __ =  LOAD/STORE (__, A)
+    MachineInstr *UseMI = MRI.use_begin(DstReg)->getParent();
+
+    for (auto &Operand : MI.operands()) {
+      // Ignore first operand (def.).
+      if (!Operand.isReg() || Operand.getReg() == DstReg)
+        continue;
+
+      const Register Reg = Operand.getReg();
+      MachineInstr *DefMI = MRI.getUniqueVRegDef(Reg);
+
+      // Ignore internal definitions (same MBB).
+      if (DefMI && DefMI->getParent() == MI.getParent())
+        continue;
+
+      if (UseMI->mayLoad())
+        LoadUses[Reg].insert(&MI);
+      else
+        UsedByStore.insert(Reg);
+    }
+  }
+
+  // The second part, filter the real useful cases,
+  // registers used in both load and stores.
+  // Then duplicate those registers.
```

Comment:
Checking, we are duplicating the components of the iterator registers and use them for load instructions. The stores will keep using the "old" registers.